src/mlpack/methods/ann/loss_functions/poisson_nll_loss.hpp

```diff
@@ -0,0 +1,142 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_HPP
+#define MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Implementation of the Poisson Negative Log Likelihood loss. This loss
+ * function expects input for each class. It also expects a class index,
+ * in the range between 1 and the number of classes, as target when calling
+ * the Forward function.
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+  typename InputDataType = arma::mat,
+  typename OutputDataType = arma::mat
+>
+class PoissonNLLLoss
+{
+ public:
+  /**
+   * Create the PoissonNLLLoss object.
+   *
+   * @param logInput If true the loss is computed as exp(input) - target*input,
+   *        if false then the loss is input - target * log(input + eps).
+   * @param full Boolean value that determines whether to include Stirling's
+   *        approximation term.
+   * @param eps A small value to prevent log(0) when logInput = False.
+   * @param reduction When true, mean loss is computed otherwise total loss.
+   */
+  PoissonNLLLoss(const bool logInput = true,
+                 const bool full = false,
+                 const double eps = 1e-08,
+                 const bool reduction = true);
+
+  /**
+   * Computes the Poisson Negative log likelihood.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param target The target vector, that contains the class index in the range
+   *        between 1 and the number of classes.
+   */
+  template<typename InputType, typename TargetType>
+  double Forward(const InputType& input, TargetType& target);
+
+  /**
+   * Ordinary feed backward pass of a neural network. The Poisson Negative Log
+   * Likelihood loss function expects the input for each class.
+   * It expects a class index, in the range between 1 and the number of classes,
+   * as target when calling the Forward function.
+   *
+   * @param input The propagated input activation.
+   * @param target The target vector, that contains the class index in the range
+   *        between 1 and the number of classes.
+   * @param output The calculated error.
+   */
+  template<typename InputType, typename TargetType, typename OutputType>
+  void Backward(const InputType& input,
+                const TargetType& target,
+                OutputType& output);
+
+  //! Get the input parameter.
+  InputDataType& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
```

Comment:
Hmm,  Hey @mrityunjay-tripathi, Could you explain why we need to do this, I might be missing something. I don't think this is needed here. 
Refer : [1](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/loss_functions/mean_squared_logarithmic_error.hpp), [2](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/loss_functions/kl_divergence.hpp), [3](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp)

Reply:
I used [negative_log_likelihood.hpp](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/loss_functions/negative_log_likelihood.hpp) as a template for this. Though this seems redundant but @zoq can give more insight into this.